Subject: [PATCH] ppro enabled patch
---
Index: subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/settings/views/ProSettingViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/settings/views/ProSettingViewModel.kt b/subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/settings/views/ProSettingViewModel.kt
--- a/subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/settings/views/ProSettingViewModel.kt	(revision 32c2ea7832b8176897acfc16c83834cdfe2c1dcb)
+++ b/subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/settings/views/ProSettingViewModel.kt	(date 1734025406188)
@@ -24,6 +24,7 @@
 import com.duckduckgo.anvil.annotations.ContributesViewModel
 import com.duckduckgo.di.scopes.ViewScope
 import com.duckduckgo.subscriptions.api.SubscriptionStatus
+import com.duckduckgo.subscriptions.api.SubscriptionStatus.EXPIRED
 import com.duckduckgo.subscriptions.api.SubscriptionStatus.UNKNOWN
 import com.duckduckgo.subscriptions.impl.SubscriptionsConstants.MONTHLY_PLAN_ROW
 import com.duckduckgo.subscriptions.impl.SubscriptionsConstants.MONTHLY_PLAN_US
@@ -93,7 +94,7 @@
                     MONTHLY_PLAN_US -> SubscriptionRegion.US
                     else -> null
                 }
-                _viewState.emit(viewState.value.copy(status = subscriptionStatus, region = region))
+                _viewState.emit(viewState.value.copy(status = UNKNOWN, region = SubscriptionRegion.US))
             }.launchIn(viewModelScope)
     }
 
Index: app/src/main/java/com/duckduckgo/app/settings/NewSettingsViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/app/src/main/java/com/duckduckgo/app/settings/NewSettingsViewModel.kt b/app/src/main/java/com/duckduckgo/app/settings/NewSettingsViewModel.kt
--- a/app/src/main/java/com/duckduckgo/app/settings/NewSettingsViewModel.kt	(revision 32c2ea7832b8176897acfc16c83834cdfe2c1dcb)
+++ b/app/src/main/java/com/duckduckgo/app/settings/NewSettingsViewModel.kt	(date 1734024761359)
@@ -173,7 +173,7 @@
                     showAutofill = autofillCapabilityChecker.canAccessCredentialManagementScreen(),
                     showSyncSetting = deviceSyncState.isFeatureEnabled(),
                     isAutoconsentEnabled = autoconsent.isSettingEnabled(),
-                    isPrivacyProEnabled = subscriptions.isEligible(),
+                    isPrivacyProEnabled = true,
                     isDuckPlayerEnabled = duckPlayer.getDuckPlayerState().let { it == ENABLED || it == DISABLED_WIH_HELP_LINK },
                     isVoiceSearchVisible = voiceSearchAvailability.isVoiceSearchSupported,
                 ),
@@ -192,7 +192,7 @@
                 val currentState = currentViewState()
                 viewState.value = currentState.copy(
                     appTrackingProtectionEnabled = isDeviceShieldEnabled,
-                    isPrivacyProEnabled = subscriptions.isEligible(),
+                    isPrivacyProEnabled = true,
                 )
                 delay(1_000)
             }
