// @duckduckgo/privacy_dashboard/shared/js/ui/templates/cta-rotation.es6.js
const bel = require('bel')
const raw = require('bel/raw')
const i18n = window.DDG.base.i18n

const ctas = {
    spread: {
        title: i18n.t('ctascreens:spreadTitle.title'),
        text: i18n.t('ctascreens:spreadText.title'),
        icon: heartArrowSvg,
        action: () => {
            return bel`<a href="https://duckduckgo.com/spread" target="_blank" class="cta__button">${i18n.t('ctascreens:spreadButton.title')}</a>`
        }
    },
    email: {
        title: i18n.t('ctascreens:emailTitle.title'),
        text: i18n.t('ctascreens:emailText.title'),
        icon: emailSvg,
        action: () => {
            return bel`<a href="https://duckduckgo.com/email" target="_blank" class="cta__button">${i18n.t('ctascreens:spreadButton.title')}</a>`
        }
    }
}

function ctaRotationView () {
    const cta = ctas[this.model.currentCta]
    if (!cta) {
        console.warn('unreachable - selected CTA not available ' + this.model.currentCta)
        return ''
    }
    return bel`
    <div class="cta-screen" data-test-id="cta-screens">
        <p class="note">${i18n.t('ctascreens:protectionsUnavailableNote.title')}</p>
        <div class="cta text--center">
            <div class="cta__icon">${cta.icon()}</div>
            <h1 class="cta__title">${cta.title}</h1>
            <h2 class="cta__text">${cta.text}</h2>
            <div class="cta__action">${cta.action()}</div>
        </div>  
    </div>
    `
}

function heartArrowSvg () {
    return raw`<svg width="100" height="100" viewBox="0 0 100 100" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M78.5138 21L71.5098 28.003V34.003L78.5138 27V21Z" fill="#C0C0C0"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M78.505 27.0034L71.51 33.9614H77.51L84.505 27.0034H78.505Z" fill="#C0C0C0"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M66.339 42.0032L63.51 39.1742L76.684 26.0012L79.512 28.8302L66.339 42.0032Z" fill="#C0C0C0"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M64.755 32.1171C58.769 31.3911 53.4 34.2001 50.505 38.6781C47.61 34.2001 42.241 31.3911 36.255 32.1171C29.715 32.9091 24.444 38.0531 23.619 44.4181C23.577 44.7471 23.546 45.0751 23.526 45.4001C23.293 49.2851 24.979 53.0491 27.799 55.8071L50.505 78.0031L73.211 55.8071C76.031 53.0491 77.717 49.2851 77.484 45.4001C77.464 45.0751 77.433 44.7471 77.391 44.4181C76.566 38.0531 71.295 32.9091 64.755 32.1171Z" fill="#DE5833"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M77.3909 44.4179C76.5659 38.0529 71.2949 32.9099 64.7549 32.1169C62.5109 31.8449 60.3539 32.0699 58.3809 32.6929C63.6639 34.3599 67.6819 38.9309 68.3929 44.4179C68.4359 44.7479 68.4669 45.0749 68.4869 45.3999C68.7189 49.2849 67.0349 53.0489 64.2129 55.8069L46.0059 73.6049L50.5049 78.0029L73.2109 55.8069C76.0319 53.0489 77.7169 49.2849 77.4839 45.3999C77.4639 45.0749 77.4329 44.7479 77.3909 44.4179Z" fill="#BC4726"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M31.51 66.0034L27.51 78.0034L39.51 74.0034L31.51 66.0034Z" fill="#C0C0C0"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M33.51 74.0034C32.998 74.0034 32.487 73.8084 32.096 73.4174C31.315 72.6364 31.315 71.3704 32.096 70.5894L44.096 58.5894C44.876 57.8084 46.144 57.8084 46.924 58.5894C47.705 59.3704 47.705 60.6364 46.924 61.4174L34.924 73.4174C34.534 73.8084 34.022 74.0034 33.51 74.0034Z" fill="#C0C0C0"/>
<path d="M86.4922 44L90.5052 41" stroke="#AAAAAA" stroke-opacity="0.6" stroke-width="3" stroke-linecap="round"/>
<path d="M88.5049 52.5H94.5099" stroke="#AAAAAA" stroke-opacity="0.6" stroke-width="3" stroke-linecap="round"/>
<path d="M14.5176 44L10.5046 41" stroke="#AAAAAA" stroke-opacity="0.6" stroke-width="3" stroke-linecap="round"/>
<path d="M12.505 52.5H6.5" stroke="#AAAAAA" stroke-opacity="0.6" stroke-width="3" stroke-linecap="round"/>
<path d="M86.4922 61.5L90.5052 64.5" stroke="#AAAAAA" stroke-opacity="0.6" stroke-width="3" stroke-linecap="round"/>
<path d="M14.5176 61.5L10.5046 64.5" stroke="#AAAAAA" stroke-opacity="0.6" stroke-width="3" stroke-linecap="round"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M27.9544 15.0818C26.3303 14.4357 24.6178 14.8261 23.4753 15.8672C22.995 14.398 21.6978 13.2137 19.9664 12.9746C18.0748 12.7131 16.2171 13.7652 15.5153 15.4882C15.4792 15.5773 15.4462 15.6669 15.4166 15.7565C15.0641 16.8282 15.2583 18.0079 15.8448 18.9894L20.5677 26.8894L28.5729 22.3469C29.5672 21.7824 30.318 20.8521 30.54 19.7459C30.5584 19.6534 30.574 19.5591 30.5865 19.4638C30.8259 17.6188 29.7289 15.7873 27.9544 15.0818Z" fill="#E2E2E2"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M77.945 75.9298C76.9297 76.1112 76.1926 76.8398 75.9437 77.7173C75.2457 77.1301 74.2269 76.94 73.2844 77.3588C72.2546 77.8163 71.6506 78.9218 71.8348 80.004C71.8444 80.0599 71.8558 80.1151 71.8688 80.1693C72.0255 80.8163 72.49 81.3473 73.0894 81.6569L77.9155 84.1482L80.5157 79.3799C80.8385 78.7875 80.9255 78.0875 80.6926 77.4638C80.673 77.4117 80.6515 77.3596 80.6282 77.3079C80.1741 76.3084 79.0542 75.7314 77.945 75.9298Z" fill="#E2E2E2"/>
</svg>

`
}

function emailSvg () {
    return raw`<svg width="76" height="78" viewBox="0 0 76 78" fill="none" xmlns="http://www.w3.org/2000/svg">
<g clip-path="url(#clip0_273_32068)">
<path d="M55.0125 26.7115C55.3701 26.3897 55.6204 25.9249 55.6204 25.3886V0.71875L41.4263 16.2L55.0125 26.7115Z" fill="#F9BE1A"/>
<path d="M14.147 0.71875V25.3886C14.147 25.9249 14.3972 26.3897 14.7548 26.7115L28.3411 16.2L14.147 0.71875Z" fill="#F9BE1A"/>
<path d="M40.9621 16.7361C40.2828 17.487 39.3175 17.8803 38.3164 17.8803H31.4517C30.4506 17.8803 29.4853 17.4512 28.806 16.7361L28.3412 16.2356L14.7549 26.7471C15.0767 27.0331 15.47 27.2119 15.9347 27.2119H53.8333C54.2981 27.2119 54.6914 27.0331 55.0132 26.7471L41.4269 16.2356L40.9621 16.7361Z" fill="#F9BE1A"/>
<path d="M55.6209 0.71875L41.4268 16.2L40.962 16.7363C40.2827 17.4871 39.3174 17.8804 38.3163 17.8804H31.4516C30.4505 17.8804 29.4852 17.4513 28.8059 16.7363L28.3411 16.2357L14.147 0.71875H55.6209Z" fill="#F9BE1A"/>
</g>
<g clip-path="url(#clip1_273_32068)">
<path d="M60.3304 38.5762C60.7824 38.1695 61.0987 37.5819 61.0987 36.9039V5.71753L43.1553 25.2881L60.3304 38.5762Z" fill="#FFCC33"/>
<path d="M8.66943 5.71753V36.9039C8.66943 37.5819 8.98582 38.1695 9.43779 38.5762L26.6129 25.2881L8.66943 5.71753Z" fill="#FFCC33"/>
<path d="M42.5679 25.966C41.7091 26.9152 40.4888 27.4123 39.2233 27.4123H30.5453C29.2798 27.4123 28.0594 26.87 27.2007 25.966L26.6131 25.3333L9.43799 38.6214C9.84477 38.983 10.3419 39.2089 10.9295 39.2089H58.839C59.4266 39.2089 59.9238 38.983 60.3306 38.6214L43.1555 25.3333L42.5679 25.966Z" fill="#FFCC33"/>
<path d="M61.0987 5.71753L43.1553 25.2881L42.5677 25.9661C41.7089 26.9152 40.4886 27.4124 39.2231 27.4124H30.5451C29.2796 27.4124 28.0592 26.87 27.2005 25.9661L26.6129 25.3333L8.66943 5.71753H61.0987Z" fill="#FFCC33"/>
</g>
<g clip-path="url(#clip2_273_32068)">
<path d="M63.5124 47.6839C64.0209 47.2263 64.3769 46.5652 64.3769 45.8025V10.7161L44.1895 32.7341L63.5124 47.6839Z" fill="#FDD20A"/>
<path d="M5.39062 10.7161V45.8025C5.39062 46.5652 5.74657 47.2263 6.25507 47.6839L25.578 32.7341L5.39062 10.7161Z" fill="#FDD20A"/>
<path d="M43.5283 33.4968C42.5622 34.5647 41.1892 35.124 39.7654 35.124H30.0023C28.5785 35.124 27.2055 34.5138 26.2394 33.4968L25.5783 32.7849L6.25537 47.7348C6.71302 48.1416 7.27237 48.3958 7.93342 48.3958H61.8343C62.4954 48.3958 63.0547 48.1416 63.5124 47.7348L44.1894 32.7849L43.5283 33.4968Z" fill="#F5B608"/>
<path d="M64.3765 10.7161L44.1891 32.7341L43.528 33.4968C42.5619 34.5647 41.189 35.124 39.7652 35.124H30.002C28.5782 35.124 27.2052 34.5138 26.2391 33.4968L25.578 32.7849L5.39062 10.7161H64.3765Z" fill="#FFDE7A"/>
</g>
<rect x="13.0835" y="47.7449" width="43.9126" height="4.87918" fill="#8CABFF"/>
<path d="M59.4354 37.9863L52.1166 52.6239H17.9624L10.6436 37.9863H0.885254V64.8218C0.885254 67.5165 3.06974 69.701 5.76443 69.701H64.3145C67.0092 69.701 69.1937 67.5165 69.1937 64.8218V37.9863H59.4354Z" fill="#668BFF"/>
<path d="M10.6436 37.9864H5.76443V23.3489H3.32484C1.97749 23.3489 0.885254 24.4411 0.885254 25.7885V40.426C0.885254 41.7733 1.97749 42.8656 3.32484 42.8656H13.0832L10.6436 37.9864Z" fill="#8CABFF"/>
<path d="M59.4357 37.9864H64.3149V23.3489H66.7544C68.1018 23.3489 69.194 24.4411 69.194 25.7885V40.426C69.194 41.7733 68.1018 42.8656 66.7544 42.8656H56.9961L59.4357 37.9864Z" fill="#8CABFF"/>
<path d="M59.1032 54.5231C60.7374 51.5626 64.9926 51.5626 66.6268 54.5231L74.8089 69.3457C76.3897 72.2095 74.3182 75.7191 71.047 75.7191H54.6829C51.4118 75.7191 49.3403 72.2095 50.9211 69.3457L59.1032 54.5231Z" fill="#FDD20A"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M62.1235 58.5312C61.6767 58.5312 61.3209 58.9053 61.3433 59.3515L61.6558 65.6015C61.6765 66.0173 62.0197 66.3438 62.436 66.3438H63.2935C63.7098 66.3438 64.0529 66.0173 64.0737 65.6015L64.3862 59.3515C64.4085 58.9053 64.0528 58.5312 63.606 58.5312H62.1235ZM62.8647 71.0312C63.7277 71.0312 64.4272 70.3317 64.4272 69.4688C64.4272 68.6058 63.7277 67.9062 62.8647 67.9062C62.0018 67.9062 61.3022 68.6058 61.3022 69.4688C61.3022 70.3317 62.0018 71.0312 62.8647 71.0312Z" fill="black" fill-opacity="0.8"/>
<defs>
<clipPath id="clip0_273_32068">
<rect x="14.147" y="0.71875" width="41.4739" height="26.4932" rx="2.99924" fill="white"/>
</clipPath>
<clipPath id="clip1_273_32068">
<rect x="8.66943" y="5.71753" width="52.4293" height="33.4915" rx="2.99924" fill="white"/>
</clipPath>
<clipPath id="clip2_273_32068">
<rect x="5.39062" y="10.7161" width="58.9859" height="37.6798" rx="2.99924" fill="white"/>
</clipPath>
</defs>
</svg>
`
}

module.exports = ctaRotationView
module.exports.ctas = ctas
