// @duckduckgo/privacy_dashboard/shared/js/ui/templates/site.es6.js
const bel = require('bel')
const raw = require('bel/raw')
const { normalizeCompanyName } = require('../models/mixins/normalize-company-name.es6.js')
const hero = require('./shared/hero.es6.js')
const toggleButton = require('./shared/toggle-button.es6.js')
const trackerNetworksIcon = require('./shared/tracker-network-icon.es6.js')
const trackerNetworksText = require('./shared/tracker-networks-text.es6.js')
const { getColorId } = require('./shared/utils.es6.js')
const i18n = window.DDG.base.i18n

const renderUpdatingSpinner = () => {
    return bel`<img src="../img/spinner.svg" style="height: 18px; width: 18px;" alt="${i18n.t('site:updatingProtectionList.title')}" />`
}

function renderHero () {
    return bel`${hero({
        status: 'hidden'
    })}`
}

module.exports = function () {
    const protectionStatus = this.model.isAllowlisted
        ? bel`${raw(i18n.t('site:protectionsDisabled.title'))}`
        : bel`${raw(i18n.t('site:protectionsEnabled.title'))}`

    const protectionToggle = this.model.tab.isPendingUpdates
        ? renderUpdatingSpinner()
        : toggleButton(!this.model.isAllowlisted, 'js-site-toggle pull-right')

    return bel`<div class="site-info site-info--main">
    ${renderHero()}
    <div class="list-wrapper">
        <ul class="default-list card-list">
            ${renderKeyInsight(this.model)}
        </ul>
    </div>
    <div class="list-wrapper">
        <ul class="default-list card-list card-list--bordered">
            <li class="js-site-tracker-networks js-site-show-page-trackers site-info__li--trackers">
                <a href="javascript:void(0)" class="link-action" role="button">
                    ${renderTrackerNetworks(this.model)}
                </a>
            </li>
            <li class="js-site-show-page-connection site-info__li--https-status border-light--top">
                <a href="javascript:void(0)" class="link-action" role="button">
                    ${renderConnection(this.model)}
                </a>
            </li>
        ${renderCookieConsentManaged(this.model)}
        </ul>
    </div>
    <div class="list-wrapper site-info__protection-wrapper">
        <ul class="default-list">
            <li class="site-info__li--toggle ${this.model.isAllowlisted ? '' : 'is-active'}">
                <p class="site-info__protection js-site-protection"><span>${protectionStatus}</span></p>
                <div class="site-info__toggle-container">${protectionToggle}</div>
            </li>
        </ul>
    </div>
    <div class="list-wrapper card-list--last">
        <ul class="default-list border-light--top">
            <li class="js-site-manage-allowlist-li site-info__li--manage-allowlist">
                ${renderManageAllowlist(this.model)}
            </li>
        </ul>
    </div>
    ${renderManagePermissions(this.model)}
</div>`
}

function renderConnection (model) {
    return bel`<div>
            <div class="site-info__trackers">
                <span class="site-info__https-status__icon is-${model.httpsState}"></span>
                <span>${model.httpsStatusText}</span>
                <span class="icon icon__arrow pull-right"></span>
            </div>
        </div>`
}

function renderTrackerNetworks (model) {
    const isActive = !model.isAllowlisted ? 'is-active' : ''

    return bel`
        <div>
            <div class="site-info__trackers">
                <span class="site-info__trackers-status__icon icon-${trackerNetworksIcon(model)}"></span>
                <span class="${isActive}">${trackerNetworksText(model, false)}</span>
                <span class="icon icon__arrow pull-right"></span>
            </div>
        </div>`
}

function renderManagePermissions (model) {
    if (!model.permissions || model.permissions.length === 0) {
        return ''
    }

    const localizedPerms = localizePermissions(model.permissions)

    return bel`<ul class="default-list">
        <li class="site-info__li--manage-permissions">
            ${localizedPerms.map(({ key: permissionId, title, permission, options }, index) => {
        return bel`<div class="site-info__page-permission ${index !== model.permissions.length - 1 ? 'border-light--bottom--inner' : ''}">
                    <label>
                        <div>
                            <div class="site-info__page-permission__icon ${permissionId}"></div>
                            ${title}
                        </div>
                        <select class="js-site-permission" name="${permissionId}">
                            ${options.map(({ id, title }) =>
        bel`<option value="${id}" ${permission === id ? 'selected' : ''}>${title}</option>`)
}
                        </select>
                    </label>
                </div>`
    })}
        </li>
    </ul>`
}

function renderManageAllowlist (model) {
    return bel`<div class="manage-allowlist">
            <a href="javascript:void(0)" class="js-site-report-broken site-info__report-broken">
                ${i18n.t('site:websiteNotWorkingQ.title')}
            </a>
        </div>`
}

function generateCompanyNamesList (model) {
    const companyNames = model.companyNames()
    return {
        companyCount: companyNames.length,
        othersCount: companyNames.length - 4,
        firstCompany: companyNames[0],
        secondCompany: companyNames[1],
        thirdCompany: companyNames[2],
        fourthCompany: companyNames[3]
    }
}

function renderCompanyIconsList (model) {
    const companyNames = model.companyNames()
    if (companyNames.length === 0) return ''

    const topCompanies = companyNames.slice(0, 4)
    const remainingCount = companyNames.length - topCompanies.length
    const remainingCountIcon = remainingCount <= 0
        ? ''
        : bel`
            <span class="site-info__tracker__icon-positioner">
                <span class="site-info__tracker__icon-wrapper site-info__tracker__icon-wrapper--count">
                    <span class="site-info__tracker__count">+${remainingCount}</span>
                </span>
            </div>
            `
    const topCompaniesIcons = topCompanies.map((name, index) => {
        const slug = normalizeCompanyName(name)
        const locationClass = index === topCompanies.length - 1 ? 'first' : 'other'

        return bel`
            <span class="site-info__tracker__icon-positioner">
                <span class="site-info__tracker__icon-wrapper site-info__tracker__icon-wrapper--${locationClass}">
                    <span class="site-info__tracker__icon ${slug[0].toUpperCase()} color-${getColorId(slug)} ${slug}"></span>
                    <span class="site-info__tracker__blocked-icon"></span>
                </span>
            </span>
            `
    })

    return bel`
            <div class="site-info__key-insight_trackers-icons">
                ${topCompaniesIcons}
                ${remainingCountIcon}
            </div>
        `
}

function renderKeyInsight (model) {
    if (model.httpsState === 'none') {
        return bel`
                <li class="site-info__li--key-insight">
                    <div class="site-info__key-insight site-info__key-insight--insecure-connection">
                        <h1>${model.tab.site.domain}</h1>
                        <div>${raw(i18n.t('site:connectionDescriptionUnencrypted.title'))}</div>
                    </div>
                </li>
            `
    }

    if (model.isAllowlisted) {
        return bel`
                <li class="site-info__li--key-insight">
                    <div class="site-info__key-insight site-info__key-insight--protections-off">
                        <h1>${model.tab.site.domain}</h1>
                        <div>
                            ${raw(i18n.t('site:protectionsDisabled.title'))}
                        </div>
                    </div>
                </li>
            `
    }

    if (model.isaMajorTrackingNetwork) {
        const company = model.tab.parentEntity

        return bel`
                <li class="site-info__li--key-insight">
                    <div class="site-info__key-insight site-info__key-insight--tracker-network">
                        <h1>${model.tab.site.domain}</h1>
                        <div>
                            ${raw(i18n.t('site:majorTrackingNetworkDesc.title', { companyDisplayName: company.displayName, companyPrevalence: Math.round(company.prevalence) }))}
                        </div>
                    </div>
                </li>
            `
    }

    if (model.trackersCount === 0) {
        return bel`
                <li class="site-info__li--key-insight">
                    <div class="site-info__key-insight site-info__key-insight--no-activity">
                        <h1>${model.tab.site.domain}</h1>
                        <div>${raw(i18n.t('site:trackerNetworksSummaryNone.title'))}</div>
                    </div>
                </li>
            `
    }

    const companyNames = model.companyNames()
    if (companyNames.length === 0) {
        return bel`
                <li class="site-info__li--key-insight">
                    <div class="site-info__key-insight site-info__key-insight--trackers-blocked">
                        <h1>${model.tab.site.domain}</h1>
                        <div><span>${raw(i18n.t('site:trackersBlockedDesc.title', generateCompanyNamesList(model)))}</span></div>
                    </div>
                </li>
            `
    }

    return bel`
            <li class="site-info__li--key-insight">
                <div class="site-info__key-insight">
                    ${renderCompanyIconsList(model)}
                    <h1>${model.tab.site.domain}</h1>
                    <div><span>${raw(i18n.t('site:trackersBlockedDesc.title', generateCompanyNamesList(model)))}</span></div>
                </div>
            </li>
        `
}

function renderCookieConsentManaged (model) {
    if (!model.tab?.consentManaged) return bel``

    const { consentManaged, optoutFailed } = model.tab.consentManaged
    if (consentManaged && !optoutFailed) {
        return bel`
            <li class="js-site-show-consent-managed site-info__li--consent-managed border-light--top">
                <div>
                    <div class="site-info__trackers">
                        <span class="site-info__https-status__icon is-secure"></span>
                        <span>${i18n.t('site:cookiesMinimized.title')}</span>
                    </div>
                </div>
            </li>
            `
    }
    return bel``
}

function localizePermissions (permissions) {
    // deep copy before mutating
    const updatedPermissions = JSON.parse(JSON.stringify(permissions))

    return updatedPermissions.map((perm) => {
        const permKey = `permissions:${perm.key}.title`
        if (i18n.exists(permKey)) {
            perm.title = i18n.t(permKey)
        }

        perm.options = perm.options.map((option) => {
            const optionKey = `permissions:${option.id}.title`
            if (i18n.exists(optionKey)) {
                option.title = i18n.t(optionKey)
            }
            return option
        })

        return perm
    })
}
