// @duckduckgo/privacy_dashboard/Gruntfile.js
module.exports = function (grunt) {
    const path = require('path')
    const through = require('through2')
    const sass = require('sass')
    require('load-grunt-tasks')(grunt)
    grunt.loadNpmTasks('grunt-execute')
    grunt.loadNpmTasks('grunt-contrib-copy')
    grunt.loadNpmTasks('grunt-contrib-watch')

    const platform = grunt.option('platform') || 'example'

    // only add `iframe.html` to 'example' platform for testing multiple screens
    const htmlFiles = ['popup.html']
    if (platform === 'example') {
        htmlFiles.push('iframe.html')
    }

    let buildPath = path.join('build', platform)

    // Output builds for apple devices to the swift package
    if (platform === 'ios' || platform === 'macos') {
        buildPath = `swift-package/Resources/${platform}/assets`
    }

    const baseFileMap = {
        ui: {
            '<%= dirs.public.js %>/base.js': ['<%= dirs.src.js %>/ui/base/index.es6.js']
        },
        sass: {
            '<%= dirs.public.css %>/base.css': ['<%= dirs.src.scss %>/base/base.scss'],
            '<%= dirs.public.css %>/popup.css': ['<%= dirs.src.scss %>/popup.scss']
        }
    }

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        dirs: {
            cache: '.cache',
            src: {
                js: 'shared/js',
                scss: 'shared/scss'
            },
            data: 'shared/data',
            public: {
                js: `${buildPath}/public/js`,
                css: `${buildPath}/public/css`
            }
        },

        browserify: {
            options: {
                transform: [
                    ['babelify'],
                    ['require-globify'],
                    [(file) => {
                        return through(function (buf, enc, next) {
                            this.push(buf.toString('utf8').replace(/\$ENVIRONMENT/g, platform))
                            next()
                        })
                    }]
                ]
            },
            ui: {
                files: baseFileMap.ui
            }
        },

        sass: {
            options: {
                implementation: sass,
                includePaths: [
                    path.resolve(process.cwd(), 'node_modules')
                ]
            },
            dist: {
                files: baseFileMap.sass
            }
        },

        copy: {
            html: {
                expand: true,
                cwd: 'shared/html',
                src: htmlFiles,
                dest: `${buildPath}/html`,
                options: {
                    process: (content) => content.replace(/\$ENVIRONMENT/g, platform)
                }
            },
            images: {
                expand: true,
                cwd: 'shared',
                src: 'img/**',
                dest: `${buildPath}/`
            }
        },
        /**
         * Run predefined tasks whenever watched files are added,
         * modified or deleted.
         */
        watch: {
            sass: {
                files: ['shared/scss/**/*'],
                tasks: ['sass']
            },
            scripts: {
                files: ['shared/js/**/*', 'shared/html/**/*'],
                tasks: ['browserify:ui', 'copy:html']
            }
        }
    })

    grunt.registerTask('build', 'Build project(s)css, templates, js', ['sass', 'browserify:ui', 'copy:html', 'copy:images'])
    grunt.registerTask('default', 'build')
    grunt.registerTask('dev', ['default', 'watch'])
}
